onnxruntime/core/providers/cuda/nn/conv.h

```diff
@@ -30,8 +31,55 @@ class CudnnConvolutionDescriptor final {
   cudnnConvolutionDescriptor_t desc_;
 };
 
+template <typename Key, typename T>
+class lru_map {
+ public:
+  lru_map(size_t max_size) : max_size_(max_size) {}
+
+  void insert(const Key& key, const T& value) {
+    auto it = handles_.find(key);
+    if (it != handles_.end()) {
+      items_.erase(it->second);
+      handles_.erase(it);
+    }
+
+    items_.emplace_front(key, value);
+    handles_.emplace(key, items_.begin());
+
+    while (size() > max_size_) {
+      handles_.erase(items_.back().first);
+      items_.pop_back();
+    }
+  }
+
+  T& at(const Key& key) {
+    auto it = handles_.find(key);
+    if (it == handles_.end()) {
+      throw std::out_of_range("There is no such key in cache");
+    }
+    items_.splice(items_.begin(), items_, it->second);
+    return it->second->second;
+  }
+
+  bool contains(const Key& key) const {
+    return handles_.find(key) != handles_.end();
+  }
+
+  size_t size() const {
+    return handles_.size();
+  }
+
+private:
+  using value_type = std::pair<Key, T>;
+  using iterator_type = typename std::list<value_type>::iterator;
+
+  size_t max_size_;
+  std::list<value_type> items_;
```

Comment:
I need to store the `T` somewhere. Currently it is stored in the `list`. It could be moved to the `(unordered_)map`, but I don't really see any advantages.

Reply:
Will this work?
std::list<Key> lru_keys_;
std::unordered_map<Key, T> items_;

Then you don't have to change items_ when adjusting lru_keys_ when touching.

The benefit of unordered_map vs. map is mainly about number of comparisons when searching: [https://www.geeksforgeeks.org/map-vs-unordered_map-c/](https://www.geeksforgeeks.org/map-vs-unordered_map-c/)

---
In reply to: [270300129](https://github.com/Microsoft/onnxruntime/pull/712#discussion_r270300129) [](ancestors = 270300129)

Reply:
Unfortunately not, because in order to move the right element to the front of the list in O(1) we need the iterator in the unordered_map. So `unordered_map<Key, pair<T, list::iterator>>` would work. But there's really not much difference I can see, because you have to touch both elements for every operation anyway.

Reply:
You are right that iterator to List needs to be stored in map. One thing can be improved is to modify the iterator value in an existing map item, instead of erase/insert in the map. Having T inside map would reduce the possible copies and memory allocation/deallocation when insert/remove items in list.

---
In reply to: [271266455](https://github.com/Microsoft/onnxruntime/pull/712#discussion_r271266455) [](ancestors = 271266455)

Reply:
Agreed, I'll change it this way